# Выбор локации для скважины

## Загрузка и подготовка данных

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

In [3]:
# данные по региону №1
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [4]:
data_1.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [5]:
data_1.duplicated('id').sum()

10

In [6]:
data_1[data_1.duplicated('id') == True]

id        f0        f1         f2     product
7530   HZww2  1.061194 -0.373969  10.430210  158.828695
41724  bxg6G -0.823752  0.546319   3.630479   93.007798
51970  A5aEY -0.180335  0.935548  -2.094773   33.020205
63593  QcMuo  0.635635 -0.473422   0.862670   64.578675
66136  74z30  1.084962 -0.312358   6.990771  127.643327
69163  AGS9W -0.933795  0.116194  -3.655896   19.230453
75715  Tdehs  0.112079  0.430296   3.218993   60.964018
90815  fiKDv  0.049883  0.841313   6.394613  137.346586
92341  TtcGQ  0.110711  1.022689   0.911381  101.318008
97785  bsk9y  0.378429  0.005837   0.160827  160.637302

In [7]:
#удалим дубликаты, сохранив в таблице последние значения, так как данные по скважине могли быть обновлены
data_1=data_1.drop_duplicates(subset='id', keep='last').reset_index(drop=True)
data_1.duplicated('id').sum()

0

In [8]:
#сохраним признаки и таргет в х_1, у_1
x_1 = data_1.drop(['id','product'], axis=1)
y_1 = data_1['product']

In [10]:
# данные по региону №2
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [11]:
data_2.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [12]:
data_2.duplicated('id').sum()

4

In [13]:
data_2[data_2.duplicated('id') == True]

id         f0        f1        f2     product
41906  LHZR0  -8.989672 -4.286607  2.009139   57.085625
82178  bfPNe  -6.202799 -4.820045  2.995107   84.038886
82873  wt4Uk  10.259972 -9.376355  4.994297  134.766305
84461  5ltQ6  18.213839  2.191999  3.993869  107.813044

In [14]:
#удалим дубликаты, сохранив в таблице последние значения, так как данные по скважине могли быть обновлены
data_2=data_2.drop_duplicates(subset='id', keep='last').reset_index(drop=True)
data_2.duplicated('id').sum()

0

In [15]:
#сохраним признаки и таргет в х_2, у_2
x_2 = data_2.drop(['id','product'], axis=1)
y_2 = data_2['product']

In [17]:
# данные по региону №3
data_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [18]:
data_3.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [19]:
data_3.duplicated('id').sum()

4

In [20]:
data_3[data_3.duplicated('id')==True]

id        f0        f1        f2     product
43233  xCHr8 -0.847066  2.101796  5.597130  184.388641
49564  VF7Jo -0.883115  0.560537  0.723601  136.233420
55967  KUPhW  1.211150  3.176408  5.543540  132.831802
95090  Vcm5J  2.587702  1.986875  2.482245   92.327572

In [21]:
#удалим дубликаты, сохранив в таблице последние значения, так как данные по скважине могли быть обновлены
data_3=data_3.drop_duplicates(subset='id', keep='last').reset_index(drop=True)
data_3.duplicated('id').sum()

0

In [22]:
#сохраним признаки и таргет в х_3, у_3
x_3 = data_3.drop(['id','product'], axis=1)
y_3 = data_3['product']

### Вывод:
- На этапе предобработки данных были удалены дубликаты id скважин из трех таблиц
- были выделены признаки для обучения (х1,х2,х3) и целевой признак(у1,у2,у3)

## Обучение и проверка модели

### Исследование региона №1

In [23]:
def lin_reg(x_train, y_train,x_valid,y_valid):
    model = LinearRegression()
    model.fit(x_train, y_train)
    y_predicted = model.predict(x_valid)
    print('Значение RMSE - ', mean_squared_error(y_valid, y_predicted,squared=False))
    print('Средний запас сырья - ', y_predicted.mean())
    return model,y_predicted

In [24]:
#разделение на тренировочную и валидационную выборки
x_train_1, x_valid_1, y_train_1, y_valid_1 = train_test_split(x_1, y_1, test_size=0.25, random_state=83749)

In [25]:
x_train_1.shape

(74992, 3)

In [26]:
x_valid_1.shape

(24998, 3)

In [27]:
#обучение модели №1
model_1, y_predicted_1 = lin_reg(x_train_1, y_train_1,x_valid_1,y_valid_1)

Значение RMSE -  37.72327730546291
Средний запас сырья -  92.48439164533188


### Исследование региона №2

In [28]:
#разделение на тренировочную и валидационную выборки
x_train_2, x_valid_2, y_train_2, y_valid_2 = train_test_split(x_2, y_2, test_size=0.25, random_state=83749)

In [29]:
x_train_2.shape

(74997, 3)

In [30]:
x_valid_2.shape

(24999, 3)

In [31]:
#обучение модели №2
model_2,y_predicted_2 = lin_reg(x_train_2, y_train_2,x_valid_2,y_valid_2)

Значение RMSE -  0.8913859604863702
Средний запас сырья -  69.10185714816008


### Исследование региона №3

In [32]:
#разделение на тренировочную и валидационную выборки
x_train_3, x_valid_3, y_train_3, y_valid_3 = train_test_split(x_3, y_3, test_size=0.25, random_state=83749)

In [33]:
x_train_3.shape

(74997, 3)

In [34]:
x_valid_3.shape

(24999, 3)

In [35]:
#обучение модели №3
model_3, y_predicted_3 = lin_reg(x_train_3, y_train_3,x_valid_3,y_valid_3)

Значение RMSE -  39.87026996897062
Средний запас сырья -  94.8417425499871


### Вывод:
 - были построены модели линейной регрессии для 3 регионов
 - для каждой модели было рассчитано RMSE (37.72, 0.89, 39.87)
 - был рассчитан средний запас сырья для каждого региона
 - Наиболее благоприятные регионы сточки зрения средних запасов сырья 1 и 3, но 2 регион имеет значительно меньшее среднеквадратическое отклонение, следовательно более узкое распределение

## Подготовка к расчёту прибыли

3.1. Все ключевые значения для расчётов сохраните в отдельных переменных.
3.2. Рассчитайте достаточный объём сырья для безубыточной разработки новой скважины. Сравните полученный объём сырья со средним запасом в каждом регионе.
3.3. Напишите выводы по этапу подготовки расчёта прибыли.

In [36]:
budget = 10*10**9
price_per_barrel = 450
price_per_product = 450*10**3
quantity = 500
best_quantity = 200

In [37]:
product_volume = budget/best_quantity/price_per_product
product_volume

111.11111111111111

### Выовод:
- Был рассчитан достаточный объём сырья для безубыточной разработки новой скважины - 111,11
- наиболее близки к данному значению средние запасы скважин из региново №1 и 3

## Расчёт прибыли и рисков 

ри разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
Бюджет на разработку скважин в регионе — 10 млрд рублей.
При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.

In [38]:
#функция для расчета прибыли
def profit (target, probabilities, best_quantity):
    probs_sorted = pd.Series(probabilities).sort_values(ascending=False)
    target=target.reset_index(drop=True)
    selected = target[probs_sorted.index][:best_quantity]
    return selected.sum()*price_per_product - budget

In [39]:
#расчет прибыли для региона №1
f"объем сырья - {profit(y_valid_1,y_predicted_1,best_quantity)/price_per_product}, прибыль - {profit(y_valid_1,y_predicted_1,best_quantity)}"


'объем сырья - 7530.241862788387, прибыль - 3388608838.254774'

In [40]:
#расчет прибыли для региона №2
f"объем сырья - {profit(y_valid_2,y_predicted_2,best_quantity)/price_per_product}, прибыль - {profit(y_valid_2,y_predicted_2,best_quantity)}"


'объем сырья - 5366.859325958913, прибыль - 2415086696.681511'

In [41]:
#расчет прибыли для региона №3
f"объем сырья - {profit(y_valid_3,y_predicted_3,best_quantity)/price_per_product}, прибыль - {profit(y_valid_3,y_predicted_3,best_quantity)}"


'объем сырья - 6360.594720253461, прибыль - 2862267624.1140575'

In [65]:
state = np.random.RandomState(12345)
def profit_riscks(y_valid,y_predicted,quantity):
    values = []
    count=0
    for i in range(1000):
        target_sample=y_valid.reset_index(drop=True).sample(n=quantity, replace=True, random_state=state)
        prob_sample = y_predicted[target_sample.index]
        cost=profit(target_sample, prob_sample, best_quantity) 
        
        if cost < 0:
            count+=1
            
        values.append(cost)
    
    values = pd.Series(values)
    print('Среднее значение:',values.mean())
    print('Нижняя граница доверительного интервала: {}, верхняя граница: {}'.format(values.quantile(.025),values.quantile(.975)))
    print('Доля, отрицательных случаев:', count/1000)
    

In [66]:
# region1
profit_riscks (y_valid_1,y_predicted_1,quantity)

Среднее значение: 435112093.09817004
Нижняя граница доверительного интервала: -36607990.427381516, верхняя граница: 914232471.7498555
Доля, отрицательных случаев: 0.038


In [67]:
# region2
profit_riscks (y_valid_2,y_predicted_2,quantity)

Среднее значение: 472834582.34909505
Нижняя граница доверительного интервала: 71067324.48485808, верхняя граница: 881540480.4005146
Доля, отрицательных случаев: 0.01


In [68]:
# region3
profit_riscks (y_valid_3,y_predicted_3,quantity)

Среднее значение: 404371589.41181207
Нижняя граница доверительного интервала: -62689473.6003538, верхняя граница: 873375183.295743
Доля, отрицательных случаев: 0.064


## Вывод:
- На этапе предобработки данных были удалены дубликаты id скважин из трех таблиц
- были выделены признаки для обучения (х1,х2,х3) и целевой признак(у1,у2,у3)
- были построены модели линейной регрессии для 3 регионов
- для каждой модели было рассчитано RMSE (37.72, 0.89, 39.87)
- был рассчитан средний запас сырья для каждого региона
- Наиболее благоприятные регионы сточки зрения средних запасов сырья 1 и 3, но 2 регион имеет значительно меньшее среднеквадратическое отклонение, следовательно более узкое распределение
- Был рассчитан достаточный объём сырья для безубыточной разработки новой скважины - 111,11
- наиболее близки к данному значению средние запасы скважин из региново №1 и 3
- Были расссчитаны прибыль и доверительный интервалы по 200 скважинам из каждого региона
- По результатам провеенного анализа наиболее благоприятным для разработки был выбран 2 регион, так как имеет наименьшее среднеквадратичное отклонение, наибольшее значение прибыли из 3 регионов и положительные границы доверительного интервала. А также наименьшую долю риска 0.01